src/ir/ir_traversing_visitor.h

```diff
@@ -24,47 +25,74 @@ namespace raksha::ir {
 // A visitor that also traverses the children of a node and allows performing
 // different actions before (PreVisit) and after (PostVisit) the children are
 // visited. Override any of the `PreVisit` and `PostVisit` methods as needed.
-template <typename Derived>
-class IRTraversingVisitor : public IRVisitor<Derived> {
+template <typename Derived, typename Result = Unit>
+class IRTraversingVisitor : public IRVisitor<Derived, Result> {
  public:
   virtual ~IRTraversingVisitor() {}
 
   // Invoked before all the children of `module` is visited.
-  virtual void PreVisit(const Module& module) {}
+  virtual Result PreVisit(const Module& module) {
+    return Result();
```

Comment:
I still think that this is not a good idea. If this visitor is used to translate from the IR to another tree-like structure, the superclass of all nodes within that structure being `Node`, it could potentially be natural to instantiate the visitor with `const Node *`. If you attempt to default-construct `Result` like this, that visitor cannot be instantiated with this class as-is.

Reply:
Reaching out via chat to try to close the loop on this, but, so that I'm not leaving your comment hanging:

One reason `PreVisit` is `virtual` is so that this is not required behavior.
Child classes can override `PreVisit` to get different behavior.

Is the problem that you're pointing to that they would need to override all three `PreVisit` functions (which can be done via templating the argument?)

You mention a pointer as an example of a case where the `Result` cannot be default  constructed, but I'd say it's also an example where the pointer cannot be safely constructed without knowledge of which type is being visitws.

Keen to hear your thoughts

Reply:
Added `GetDefaultValue` and set `PreVisit` to return it by default.
I'm not sure this is the right call (as not all types *can* have a meaningful default value), but it's not a deal breaker :)